File: bindings/java/umicp-examples/src/main/java/com/hivellm/umicp/examples/CompressionExample.java
package com.hivellm.umicp.examples;

import com.hivellm.umicp.compression.Compression;
import com.hivellm.umicp.types.CompressionType;
import com.hivellm.umicp.types.UMICPException;

import java.nio.charset.StandardCharsets;
import java.util.Random;

/**
 * Example demonstrating UMICP Compression
 */
public class CompressionExample {
    public static void main(String[] args) {
        System.out.println("=== UMICP Compression Example ===\n");

        try {
            // Example 1: Compress and decompress text
            System.out.println("1. Text Compression (GZIP):");
            String text = "Hello, UMICP! This is a test message for compression. " +
                         "Compression is very useful for reducing network bandwidth.";
            byte[] originalData = text.getBytes(StandardCharsets.UTF_8);

            System.out.println("   Original text: " + text);
            System.out.println("   Original size: " + originalData.length + " bytes");

            byte[] compressed = Compression.compress(originalData, CompressionType.GZIP);
            System.out.println("   Compressed size: " + compressed.length + " bytes");
            System.out.println("   Compression ratio: " +
                String.format("%.2f", Compression.getCompressionRatio(originalData.length, compressed.length)) + "x");
            System.out.println("   Beneficial: " +
                Compression.isBeneficial(originalData.length, compressed.length));

            byte[] decompressed = Compression.decompress(compressed, CompressionType.GZIP);
            String result = new String(decompressed, StandardCharsets.UTF_8);

            System.out.println("   Decompressed: " + result);
            System.out.println("   ✓ Compression successful!");
            System.out.println();

            // Example 2: Large repetitive data
            System.out.println("2. Large Repetitive Data (GZIP):");
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < 100; i++) {
                sb.append("This is line ").append(i).append(" of repetitive test data.\n");
            }
            byte[] largeData = sb.toString().getBytes(StandardCharsets.UTF_8);

            System.out.println("   Original size: " + largeData.length + " bytes");

            byte[] largeCompressed = Compression.compress(largeData, CompressionType.GZIP);
            System.out.println("   Compressed size: " + largeCompressed.length + " bytes");
            System.out.println("   Compression ratio: " +
                String.format("%.2f", Compression.getCompressionRatio(largeData.length, largeCompressed.length)) + "x");
            System.out.println("   Savings: " +
                String.format("%.1f%%", (1.0 - (double)largeCompressed.length/largeData.length) * 100));
            System.out.println();

            // Example 3: GZIP vs DEFLATE
            System.out.println("3. GZIP vs DEFLATE:");
            byte[] testData = "Test data for comparing compression algorithms".getBytes(StandardCharsets.UTF_8);

            byte[] gzipCompressed = Compression.compress(testData, CompressionType.GZIP);
            byte[] deflateCompressed = Compression.compress(testData, CompressionType.DEFLATE);

            System.out.println("   Original: " + testData.length + " bytes");
            System.out.println("   GZIP: " + gzipCompressed.length + " bytes");
            System.out.println("   DEFLATE: " + deflateCompressed.length + " bytes");
            System.out.println("   DEFLATE is smaller: " + (deflateCompressed.length < gzipCompressed.length));
            System.out.println();

            // Example 4: Random data (doesn't compress well)
            System.out.println("4. Random Data Compression:");
            byte[] randomData = new byte[1000];
            new Random(42).nextBytes(randomData);

            System.out.println("   Original size: " + randomData.length + " bytes");

            byte[] randomCompressed = Compression.compress(randomData, CompressionType.GZIP);
            System.out.println("   Compressed size: " + randomCompressed.length + " bytes");
            System.out.println("   Compression ratio: " +
                String.format("%.2f", Compression.getCompressionRatio(randomData.length, randomCompressed.length)) + "x");
            System.out.println("   Beneficial: " +
                Compression.isBeneficial(randomData.length, randomCompressed.length));
            System.out.println("   Note: Random data compresses poorly");
            System.out.println();

            // Example 5: No compression
            System.out.println("5. No Compression:");
            byte[] noneData = "test".getBytes();
            byte[] noneResult = Compression.compress(noneData, CompressionType.NONE);

            System.out.println("   Original size: " + noneData.length);
            System.out.println("   Result size: " + noneResult.length);
            System.out.println("   Same data: " + (noneData == noneResult));
            System.out.println();

            // Example 6: Special characters
            System.out.println("6. Unicode and Special Characters:");
            String unicode = "Special chars: 你好世界 🌍 émojis 😀 symbols: @#$%^&*()";
            byte[] unicodeData = unicode.getBytes(StandardCharsets.UTF_8);

            System.out.println("   Original: " + unicode);
            System.out.println("   Original size: " + unicodeData.length + " bytes");

            byte[] unicodeCompressed = Compression.compress(unicodeData, CompressionType.GZIP);
            byte[] unicodeDecompressed = Compression.decompress(unicodeCompressed, CompressionType.GZIP);
            String unicodeResult = new String(unicodeDecompressed, StandardCharsets.UTF_8);

            System.out.println("   Compressed size: " + unicodeCompressed.length + " bytes");
            System.out.println("   Decompressed: " + unicodeResult);
            System.out.println("   Match: " + unicode.equals(unicodeResult));
            System.out.println();

            System.out.println("=== Compression Example Complete ===");

        } catch (UMICPException e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: bindings/kotlin/src/main/kotlin/com/hivellm/umicp/examples/CompressionExample.kt
package com.hivellm.umicp.examples

import com.hivellm.umicp.compression.Compression
import com.hivellm.umicp.compression.CompressionAlgorithm
import com.hivellm.umicp.compression.compress
import com.hivellm.umicp.compression.decompress
import com.hivellm.umicp.core.Envelope
import com.hivellm.umicp.types.OperationType

/**
 * Compression Example
 *
 * Demonstrates:
 * - GZIP compression
 * - DEFLATE compression
 * - Compression statistics
 * - Extension functions
 */
fun main() {
    println("=== UMICP Kotlin SDK - Compression Example ===\n")

    // Sample data
    println("1. Basic Compression:")
    val originalText = "Hello, UMICP! This is a test message that demonstrates compression capabilities.".repeat(50)
    val originalData = originalText.toByteArray()

    println("  Original size: ${originalData.size} bytes")
    println("  Original text preview: ${originalText.take(80)}...")
    println()

    // GZIP Compression
    println("2. GZIP Compression:")
    val gzipCompressed = Compression.compress(originalData, CompressionAlgorithm.GZIP)
    val gzipDecompressed = Compression.decompress(gzipCompressed, CompressionAlgorithm.GZIP)

    println("  Compressed size: ${gzipCompressed.size} bytes")
    println("  Decompressed size: ${gzipDecompressed.size} bytes")
    println("  Ratio: ${"%.2f".format(Compression.getCompressionRatio(originalData.size, gzipCompressed.size))}")
    println("  Savings: ${"%.1f".format(Compression.getSavings(originalData.size, gzipCompressed.size))}%")
    println("  Match: ${String(gzipDecompressed, Charsets.UTF_8) == originalText}")
    println()

    // DEFLATE Compression
    println("3. DEFLATE Compression:")
    val deflateCompressed = Compression.compress(originalData, CompressionAlgorithm.DEFLATE)
    val deflateDecompressed = Compression.decompress(deflateCompressed, CompressionAlgorithm.DEFLATE)

    println("  Compressed size: ${deflateCompressed.size} bytes")
    println("  Savings: ${"%.1f".format(Compression.getSavings(originalData.size, deflateCompressed.size))}%")
    println("  Match: ${String(deflateDecompressed, Charsets.UTF_8) == originalText}")
    println()

    // Compression Levels
    println("4. Compression Levels (GZIP):")
    for (level in listOf(1, 6, 9)) {
        val compressed = Compression.compress(originalData, CompressionAlgorithm.GZIP, level)
        println("  Level $level: ${compressed.size} bytes (${("%.1f".format(Compression.getSavings(originalData.size, compressed.size)))}% saved)")
    }
    println()

    // String Compression
    println("5. String Compression:")
    val text = "Quick compression test! ".repeat(100)
    val compressedString = Compression.compressString(text, CompressionAlgorithm.GZIP)
    val decompressedString = Compression.decompressString(compressedString, CompressionAlgorithm.GZIP)

    println("  Original: ${text.length} chars")
    println("  Compressed: ${compressedString.size} bytes")
    println("  Match: ${decompressedString == text}")
    println()

    // Extension Functions
    println("6. Extension Functions:")
    val data = "Test data ".repeat(200).toByteArray()
    val compressed = data.compress(CompressionAlgorithm.GZIP)
    val decompressed = compressed.decompress(CompressionAlgorithm.GZIP)

    println("  Original: ${data.size} bytes")
    println("  Compressed: ${compressed.size} bytes")
    println("  Decompressed: ${decompressed.size} bytes")
    println("  Match: ${decompressed.contentEquals(data)}")
    println()

    // Compressing Envelopes
    println("7. Compressing Envelopes:")
    val envelope = Envelope.build {
        from("client-001")
        to("server-001")
        operation(OperationType.DATA)
        capability("data", "A".repeat(1000))
    }

    val envelopeJson = envelope.serialize()
    val envelopeCompressed = envelopeJson.compress(CompressionAlgorithm.GZIP)

    println("  JSON size: ${envelopeJson.length} bytes")
    println("  Compressed: ${envelopeCompressed.size} bytes")
    println("  Savings: ${"%.1f".format(Compression.getSavings(envelopeJson.length, envelopeCompressed.size))}%")
    println()

    // Highly Compressible Data
    println("8. Highly Repetitive Data:")
    val repetitive = "A".repeat(10000).toByteArray()
    val highlyCompressed = repetitive.compress(CompressionAlgorithm.GZIP)
    val savings = Compression.getSavings(repetitive.size, highlyCompressed.size)

    println("  Original: ${repetitive.size} bytes (all 'A's)")
    println("  Compressed: ${highlyCompressed.size} bytes")
    println("  Savings: ${"%.2f".format(savings)}%")
    println()

    // LZ4 Compression
    println("9. LZ4 Compression (Fast):")
    val lz4Data = "Fast compression test! ".repeat(100).toByteArray()

    val lz4Start = System.nanoTime()
    val lz4Compressed = lz4Data.compress(CompressionAlgorithm.LZ4)
    val lz4Time = (System.nanoTime() - lz4Start) / 1_000_000.0

    val lz4Decompressed = lz4Compressed.decompress(CompressionAlgorithm.LZ4)

    println("  Original: ${lz4Data.size} bytes")
    println("  Compressed: ${lz4Compressed.size} bytes")
    println("  Savings: ${"%.1f".format(Compression.getSavings(lz4Data.size, lz4Compressed.size))}%")
    println("  Time: ${"%.3f".format(lz4Time)}ms")
    println("  Match: ${lz4Decompressed.contentEquals(lz4Data)}")
    println()

    // LZ4 High Compression
    println("10. LZ4 High Compression:")
    val lz4HcCompressed = lz4Data.compress(CompressionAlgorithm.LZ4_HC, 9)
    val lz4HcDecompressed = lz4HcCompressed.decompress(CompressionAlgorithm.LZ4_HC)

    println("  LZ4:    ${lz4Compressed.size} bytes")
    println("  LZ4_HC: ${lz4HcCompressed.size} bytes")
    println("  Improvement: ${lz4Compressed.size - lz4HcCompressed.size} bytes")
    println("  Match: ${lz4HcDecompressed.contentEquals(lz4Data)}")
    println()

    // Comparison
    println("11. Algorithm Comparison:")
    val testData = "Sample data for comparison. ".repeat(100).toByteArray()

    val gzipSize = testData.compress(CompressionAlgorithm.GZIP).size
    val deflateSize = testData.compress(CompressionAlgorithm.DEFLATE).size
    val lz4Size = testData.compress(CompressionAlgorithm.LZ4).size
    val lz4HcSize = testData.compress(CompressionAlgorithm.LZ4_HC, 9).size

    println("  Original: ${testData.size} bytes")
    println("  GZIP:     $gzipSize bytes (${"%.1f".format(Compression.getSavings(testData.size, gzipSize))}% saved)")
    println("  DEFLATE:  $deflateSize bytes (${"%.1f".format(Compression.getSavings(testData.size, deflateSize))}% saved)")
    println("  LZ4:      $lz4Size bytes (${"%.1f".format(Compression.getSavings(testData.size, lz4Size))}% saved) [FASTEST]")
    println("  LZ4_HC:   $lz4HcSize bytes (${"%.1f".format(Compression.getSavings(testData.size, lz4HcSize))}% saved)")
    println()

    // Speed Comparison
    println("12. Speed Comparison (1MB data):")
    val bigData = "X".repeat(1024 * 1024).toByteArray()

    println("  Testing GZIP...")
    val gzipSpeedStart = System.nanoTime()
    bigData.compress(CompressionAlgorithm.GZIP)
    val gzipSpeed = (System.nanoTime() - gzipSpeedStart) / 1_000_000.0

    println("  Testing LZ4...")
    val lz4SpeedStart = System.nanoTime()
    bigData.compress(CompressionAlgorithm.LZ4)
    val lz4Speed = (System.nanoTime() - lz4SpeedStart) / 1_000_000.0

    println("  GZIP: ${"%.1f".format(gzipSpeed)}ms")
    println("  LZ4:  ${"%.1f".format(lz4Speed)}ms")
    println("  LZ4 is ${"%.1f".format(gzipSpeed / lz4Speed)}x faster!")
    println()

    // Use Case: Large Message Transport
    println("13. Use Case - Large Message Transport:")
    val largeMessage = """
        {
            "model": "gpt-4",
            "messages": [
                {"role": "user", "content": "${"This is a large message. ".repeat(500)}"}
            ]
        }
    """.trimIndent()

    val original = largeMessage.toByteArray()

    println("  Original message: ${original.size} bytes")
    println()
    println("  Testing compression algorithms:")

    val gzipComp = original.compress(CompressionAlgorithm.GZIP)
    println("    GZIP:   ${gzipComp.size} bytes (${"%.1f".format(Compression.getSavings(original.size, gzipComp.size))}% saved)")

    val lz4Comp = original.compress(CompressionAlgorithm.LZ4)
    println("    LZ4:    ${lz4Comp.size} bytes (${"%.1f".format(Compression.getSavings(original.size, lz4Comp.size))}% saved) [FASTEST]")

    println()
    println("  Recommendation:")
    println("    - Use LZ4 for real-time applications (fastest)")
    println("    - Use GZIP for bandwidth-constrained networks (best compression)")
    println("    - Use LZ4_HC for balanced approach")
    println()

    println("=== Example Complete ===")
}

